waku/waku_archive/retention_policy/retention_policy_size.nim

```diff
@@ -0,0 +1,98 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/times,
+  stew/results,
+  chronicles,
+  chronos,
+  os
+import
+  ../driver,
+  ../retention_policy
+
+logScope:
+  topics = "waku archive retention_policy"
+
+# default size is 30 Gb
+const DefaultRetentionSize*: float = 30_720
+
+# to remove 20% of the outdated data from database
+const DeleteLimit = 0.80
+
+type
+  # SizeRetentionPolicy implements auto delete as follows:
+  # - sizeLimit is the size in megabytes (Mbs) the database can grow upto
+  # to reduce the size of the databases, remove the rows/number-of-messages
+  # DeleteLimit is the total number of messages to delete beyond this limit
+  # when the database size crosses the sizeLimit, then only a fraction of messages are kept,
+  # rest of the outdated message are deleted using deleteOldestMessagesNotWithinLimit(),
+  # upon deletion process the fragmented space is retrieve back using Vacuum process. 
+  SizeRetentionPolicy* = ref object of RetentionPolicy
+      sizeLimit: float
+
+proc init*(T: type SizeRetentionPolicy, size=DefaultRetentionSize): T =
+  SizeRetentionPolicy(
+    sizeLimit: size
+  )
+
+method execute*(p: SizeRetentionPolicy,
+                driver: ArchiveDriver):
+                Future[RetentionPolicyResult[void]] {.async.} =
+  ## when db size overshoots the database limit, shread 20% of outdated messages 
+  
+  # to get the size of the database, pageCount and PageSize is required
+  # get page count in "messages" database
+  var pageCount = (await driver.getPagesCount()).valueOr:
+    return err("failed to get Pages count: " & $error)
+
+  # get page size of database
+  var pageSizeRes = await driver.getPagesSize()
+  var pageSize: int64 = int64(pageSizeRes.valueOr(0) div 1024)
+
+  if pageSize == 0:
+    return err("failed to get Page size: " & pageSizeRes.error)
+
+  # database size in megabytes (Mb)
+  var totalSizeOfDB: float = float(pageSize * pageCount)/1024.0
+
+  # check if current databse size crosses the db size limit
+  if totalSizeOfDB < p.sizeLimit:
+    return ok()
+
+  # keep deleting until the current db size falls within size limit 
+  while totalSizeOfDB > p.sizeLimit:
```

Comment:
This logic looks great to me! However, given the possible issue of stepping into an infinite loop, I think we should have a PR in the near future where we prevent that possible blocking issue. For example, if the measured `totalSizeOfDB` doesn't change in two consecutive iterations, then we should break the loop by returning the appropriate error.

Reply:
yes, I do not like loops in the code, what I can imagine we can do is, since we have a max size/threshold of the database, we can have a lower threshold of the database as well, then inside the loop, we introduce a notation, i.e. max number of times the loop should run so that the size of the database will reduce to the lowest,  

$$ n > \frac{\log(\text{lowerThresholdInMbs}/\text{maxDbSizeOrCurrentDbSize})}{\log(\text{reducedToUponEachIteration})} $$

Here: 
lowerThresholdInMbs is the lowest possible database size possible
maxDbSizeOrCurrentDbSize is the maximum database size possible
reducedToUponEachIteration is the fraction of size the database reduced to upon each iteration

For eg., in the case of 30720 MBs(30 GB) as max size, 0.1 MBs as a lower threshold, and 0.80 as a fraction of the database retained upon each iteration. We get 57 iterations. SO if we use this then there is a guarantee that the loop will not cross 57 times at max.

This way there is a good chance that the forever loop is not encountered.
Post this we can affirm a warning/error if the database size is still not less than the maximum permitted size.

WDYT? @Ivansete-status @alrevuelta

Reply:
I vote to follow the simplest approach. However it is fine for this PR for now and we can revisit that in a separate PR in the near future.